sharktank/sharktank/tools/tuner/candidate_gen.py

```diff
@@ -0,0 +1,1386 @@
+        get_transform_function_batch_matmul(
+            problem_size, tile_dims, f"match_op", configuration
+        ),
+        "  ",
+    )
+    return modified, embeddable
+
+
+def apply_params_batch_mmt(
+    problem_size: ProblemSize, template: list[str], configuration: Configuration
+) -> tuple[str, str]:
+    M, N, K = problem_size.MNK
+    B = problem_size.matmul_size.B
+    modified = indent(
+        get_transform_function_batch_mmt(
+            problem_size, f"match_batch_mmt_{B}x{M}x{N}x{K}", configuration
+        ),
+        "//   ",
+    )
+    modified += apply_configuration(
+        template, configuration, get_batch_mmt_tile_sizes(configuration)
+    )
+
+    embeddable = indent(
+        get_transform_function_batch_mmt(problem_size, f"match_op", configuration),
+        "  ",
+    )
+    return modified, embeddable
+
+
+def apply_params_broadcast_rhs_mmt(
+    problem_size: ProblemSize, template: list[str], configuration: Configuration
+) -> tuple[str, str]:
+    M, N, K = problem_size.MNK
+    modified = indent(
+        get_transform_function_broadcast_rhs_mmt(
+            problem_size, f"match_broadcast_rhs_mmt_Bx{M}x{N}x{K}", configuration
+        ),
+        "//   ",
+    )
+    modified += apply_configuration(
+        template, configuration, get_batch_mmt_tile_sizes(configuration)
+    )
+
+    embeddable = indent(
+        get_transform_function_broadcast_rhs_mmt(
+            problem_size, f"match_op", configuration
+        ),
+        "  ",
+    )
+    return modified, embeddable
+
+
+def parse_tensor_type(tensor_type: str) -> ShapedType:
+    shape_match = re.search(MlirRegex.tensor_type, tensor_type)
+    assert shape_match
+
+    shape_str = shape_match.group(1)
+    dims_and_elem = shape_str.split("x")
+    dims = [int(x) for x in dims_and_elem[:-1]]
+    elem = dims_and_elem[-1]
+    str_to_elem_ty = {x.name: x for x in ElementType}
+    return ShapedType(dims, str_to_elem_ty[elem])
+
+
+def get_shapes_mmt(template: list[str]) -> ProblemSize:
+    for line in template:
+        if "linalg.generic" not in line:
+            continue
+        if r'iterator_types = ["parallel", "parallel", "reduction"]' not in line:
+            continue
+        # ins(%13, %14 : tensor<2048x1280xf16>, tensor<1280x1280xf16>) outs(%19 : tensor<2048x1280xf32>)
+        mmt_re = rf"{MlirRegex.dps_ins_two_args()}\s+{MlirRegex.dps_outs_one_arg()}"
+        dps = re.search(mmt_re, line)
+        if dps is None:
+            continue
+
+        lhs_tensor_type = dps.group("LHS")
+        rhs_tensor_type = dps.group("RHS")
+        lhs_shaped_type = parse_tensor_type(lhs_tensor_type)
+        assert lhs_shaped_type.rank() == 2
+        lhs_M, lhs_K = lhs_shaped_type.shape
+
+        rhs_shaped_type = parse_tensor_type(rhs_tensor_type)
+        assert rhs_shaped_type.rank() == 2
+        rhs_N, rhs_K = rhs_shaped_type.shape
+
+        assert lhs_shaped_type.element_type == rhs_shaped_type.element_type
+        assert lhs_K == rhs_K
```

Comment:
Are these asserts possible to hit on valid programs? If so, this should probably be wrapped in some error handling code that prints a more detailed error message ("unimplemented feature", "unexpected input program structure", etc.)

Reply:
No, they can't hit on valid mlir but can triger on invalid mlir. Invalid mlir should fail to parse earlier on and this won't fire there either.

Reply:
leave it unchanged for now